File: solutions/cpp/queen-attack/1/queen_attack.cpp
#include "queen_attack.h"

namespace queen_attack {
    bool chess_board::is_valid(std::pair<int, int> position) {
        return !(position.first < 0 || position.first > 7 ||
            position.second < 0 || position.second > 7);
    }

    std::pair<int, int> chess_board::white() const {
        return white_;
    }

    std::pair<int, int> chess_board::black() const {
        return black_;
    }

    bool chess_board::can_attack() const {
        return white_.first - white_.second == 
            black_.first - black_.second ||

            white_.first + white_.second ==
            black_.first + black_.second ||

            white_.first == black_.first ||

            white_.second == black_.second;
    }
}

File: solutions/clojure/queen-attack/1/src/queen_attack.clj
(ns queen-attack)

(defn string-to-board [s]
  (let [result 
        (->> (clojure.string/split s #"\n")
             (map #(filter (partial not= \space) %))
             (map-indexed 
              (fn [row row-itm]
                (map-indexed 
                 (fn [col col-itm]
                   [col-itm [row col]])
                 row-itm)))
             (apply concat)
             (filter #(not= \_ (first %)))
             (apply concat)
             (map #(case %
                     \W :w
                     \B :b
                     %)))]
    (if (empty? result)
      {}
      (apply (partial assoc {}) result))))

(def board-size 8)

(defn board-string [board]
  (let [{white :w
         black :b} board]
    (str 
     (->> (map-indexed (fn [row row-itm] 
                         (map-indexed 
                          (fn [col col-itm] 
                            (cond
                              (= white [row col]) \W
                              (= black [row col]) \B
                              :else col-itm))
                          (repeat board-size "_"))) 
                       (range board-size))
          (map (partial clojure.string/join " "))
          (clojure.string/join "\n"))
     "\n")))


(defn can-attack [board]
  (let [{white :w
         black :b} board
        [wx wy] white
        [bx by] black]
    (cond
      (= wx bx) true
      (= wy by) true
      (= (apply + white) (apply + black)) true
      (= (apply - white) (apply - black)) true
      :else false)))
